File: client/semur/lib/accessors/users/user_chat_accessor.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:semur/accessors/users/user_accessor.dart';
import 'package:semur/config/application.dart';
import 'package:semur/models.pb/chats/chat.pb.dart';
import 'package:semur/transformers/users/chat_firebase_transformer.dart';

enum UserChatAccessorQueryKey { ALL }

Map<UserChatAccessorQueryKey, List<String>> userChatAccessorQueryArguments = {
  UserChatAccessorQueryKey.ALL: ["userId"],
};

class UserChatAccessor {
  static String firebaseCollectionName = "chats";

  final FirebaseFirestore firestore;
  UserChatAccessor({required this.firestore});

  Query buildQuery({
    required UserChatAccessorQueryKey queryKey,
    Map<String, dynamic> arguments = const {},
  }) {
    // Validate required arguments
    List<String> requiredArgs = userChatAccessorQueryArguments[queryKey] ?? [];
    for (String arg in requiredArgs) {
      if (!arguments.containsKey(arg)) {
        throw Exception("Missing required argument '$arg' for query $queryKey");
      }
    }

    // Build query based on key
    switch (queryKey) {
      case UserChatAccessorQueryKey.ALL:
        return firestore
            .collection(UserAccessor.firebaseCollectionName)
            .doc(arguments['userId'])
            .collection(firebaseCollectionName);
    }
  }

  Future<List<ChatSession>> getQuery({
    required PermissionRoles callerRole,
    required UserChatAccessorQueryKey queryKey,
    Map<String, dynamic> arguments = const {},
    int? limit,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];

    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }

    try {
      Query query = buildQuery(queryKey: queryKey, arguments: arguments);
      if (limit != null) {
        query = query.limit(limit);
      }
      QuerySnapshot snapshot = await query.get();
      return snapshot.docs
          .map((doc) => ChatFirebaseTransformer.fromFirebase(doc))
          .toList();
    } catch (e) {
      throw Exception("Error fetching chats: ${e.toString()}");
    }
  }

  Future<int> countQuery({
    required PermissionRoles callerRole,
    required UserChatAccessorQueryKey queryKey,
    Map<String, dynamic> arguments = const {},
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];

    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }

    try {
      AggregateQuerySnapshot snapshot =
          await buildQuery(
            queryKey: queryKey,
            arguments: arguments,
          ).count().get();
      return snapshot.count ?? 0;
    } catch (e) {
      throw Exception("Error fetching users: ${e.toString()}");
    }
  }

  Future<ChatSession> get({
    required String userId,
    required String chatId,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }
    try {
      DocumentSnapshot<Map<String, dynamic>> snapshot =
          await firestore
              .collection(UserAccessor.firebaseCollectionName)
              .doc(userId)
              .collection(firebaseCollectionName)
              .doc(chatId)
              .get();

      if (!snapshot.exists) {
        throw Exception("Chat not found.");
      }
      return ChatFirebaseTransformer.fromFirebase(snapshot);
    } catch (e) {
      throw Exception("Error fetching chat: ${e.toString()}");
    }
  }

  Future<void> set({
    required ChatSession chat,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to save this data.");
    }
    try {
      await firestore
          .collection(UserAccessor.firebaseCollectionName)
          .doc(chat.userId)
          .collection(firebaseCollectionName)
          .doc(chat.id)
          .set(ChatFirebaseTransformer.toJson(chat), SetOptions(merge: true));
    } catch (e) {
      throw Exception("Error saving chat: ${e.toString()}");
    }
  }

  Future<void> delete({
    required String userId,
    required String chatId,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [PermissionRoles.admin];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to delete this data.");
    }
    try {
      await firestore
          .collection(UserAccessor.firebaseCollectionName)
          .doc(userId)
          .collection(firebaseCollectionName)
          .doc(chatId)
          .delete();
    } catch (e) {
      throw Exception("Error deleting chat: ${e.toString()}");
    }
  }

  Future<void> update({
    required ChatSession chat,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to update this data.");
    }
    try {
      await firestore
          .collection(UserAccessor.firebaseCollectionName)
          .doc(chat.userId)
          .collection(firebaseCollectionName)
          .doc(chat.id)
          .update(ChatFirebaseTransformer.toJson(chat));
    } catch (e) {
      throw Exception("Error updating chat: ${e.toString()}");
    }
  }

  Future<void> updateCustomFields({
    required String userId,
    required String chatId,
    required Map<String, dynamic> updatedData,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to update this data.");
    }
    try {
      await firestore
          .collection(UserAccessor.firebaseCollectionName)
          .doc(userId)
          .collection(firebaseCollectionName)
          .doc(chatId)
          .update(updatedData);
    } catch (e) {
      throw Exception("Error updating chat: ${e.toString()}");
    }
  }

  Future<bool> exists({
    required String userId,
    required String chatId,
    required PermissionRoles callerRole,
  }) async {
    List<PermissionRoles> allowedRoles = [
      PermissionRoles.admin,
      PermissionRoles.hostUser,
      PermissionRoles.businessUser,
      PermissionRoles.user,
    ];
    if (!allowedRoles.contains(callerRole)) {
      throw Exception("You do not have permission to access this data.");
    }
    try {
      DocumentSnapshot<Map<String, dynamic>> snapshot =
          await firestore
              .collection(UserAccessor.firebaseCollectionName)
              .doc(userId)
              .collection(firebaseCollectionName)
              .doc(chatId)
              .get();

      return snapshot.exists;
    } catch (e) {
      throw Exception("Error checking chat existence: ${e.toString()}");
    }
  }
}

File: server/functions/src/accessors/users/user_chat_accessor.ts
import {SemurEngineConfig} from "../../config";
import {SemurEngineErrorBuilder} from "../../semur_engine/semur_engine";
import {ChatMessage, ChatSession} from "../../models.pb/chats/chat";
import {UserAccessor} from "./user_accessor";

export enum UserChatQueryKey {
    All = "all",
}

export enum UserChatMessageQueryKey {
    All = "all",
}

// User query arguments from query key
export interface UserChatQueryArgs {
    [UserChatQueryKey.All]: Record<string, never>;
}

export interface UserChatMessageQueryArgs {
    [UserChatMessageQueryKey.All]: {
        userId: string;
        chatId: string;
        limit?: number;
    };
}

export class UserChatAccessor {
    static firestoreCollectionName = "chats";
    static firestoreUsersCollection = SemurEngineConfig.db.collection(UserAccessor.firestoreCollectionName);
    static firestoreMessagesSubcollectionName = "messages";

    static async exists(userId: string, chatId: string): Promise<boolean> {
        const snapshot =
            await this.firestoreUsersCollection
                .doc(userId).collection(this.firestoreCollectionName)
                .doc(chatId).get();
        return snapshot.exists;
    }

    static async get(userId: string, chatId: string): Promise<ChatSession> {
        const snapshot = await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreCollectionName)
            .doc(chatId).get();
        if (!snapshot.exists) {
            throw SemurEngineErrorBuilder.notFound("Chat session not found");
        }
        return ChatSession.fromJSON(snapshot.data() as Record<string, unknown>);
    }

    static async set(userId: string, chat: ChatSession): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreCollectionName)
            .doc(chat.id).set({...chat});
    }

    static async delete(userId: string, chatId: string): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreCollectionName)
            .doc(chatId).delete();
    }

    static async update(userId: string, chat: ChatSession): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreCollectionName)
            .doc(chat.id).update({...chat});
    }

    static async updateCustomFields(userId: string, chatId: string, data: { [key: string]: unknown }): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreCollectionName)
            .doc(chatId).update(data);
    }

    static async query(queryKey: UserChatQueryKey, args: UserChatQueryArgs[UserChatQueryKey] = {}): Promise<ChatSession[]> {
        const query = await this.buildQuery(queryKey, args);
        const querySnapshot = await query.get();
        return querySnapshot.docs.map((doc) => ChatSession.fromJSON(doc.data() as Record<string, unknown>));
    }

    // Build query given query key and arguments
    static async buildQuery(queryKey: UserChatQueryKey, args: UserChatQueryArgs[UserChatQueryKey] = {}): Promise<FirebaseFirestore.Query> {
        if (!this.firestoreUsersCollection) {
            throw SemurEngineErrorBuilder.internalError("Firebase users collection is not defined");
        }
        switch (queryKey) {
            case UserChatQueryKey.All:
                return this.firestoreUsersCollection
                    .doc(args.userId).collection(this.firestoreCollectionName);
            default:
                throw SemurEngineErrorBuilder.internalError(`Unknown query key: ${queryKey}`);
        }
    }

    // Messages
    static async messageExists(userId: string, chatId: string, messageId: string): Promise<boolean> {
        const snapshot = await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreCollectionName)
            .doc(chatId).collection(this.firestoreMessagesSubcollectionName)
            .doc(messageId).get();
        return snapshot.exists;
    }

    static async messageGet(userId: string, chatId: string, messageId: string): Promise<ChatSession> {
        const snapshot = await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreCollectionName)
            .doc(chatId).collection(this.firestoreMessagesSubcollectionName)
            .doc(messageId).get();
        if (!snapshot.exists) {
            throw SemurEngineErrorBuilder.notFound("Chat message not found");
        }
        return ChatSession.fromJSON(snapshot.data() as Record<string, unknown>);
    }

    static async messageSet(userId: string, chatId: string, message: ChatMessage): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreCollectionName)
            .doc(chatId).collection(this.firestoreMessagesSubcollectionName)
            .doc(message.id.toString()).set({...message});
    }

    static async messageUpdate(userId: string, chatId: string, message: ChatMessage): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreCollectionName)
            .doc(chatId).collection(this.firestoreMessagesSubcollectionName)
            .doc(message.id.toString()).update({...message});
    }

    static async messageDelete(userId: string, chatId: string, messageId: string): Promise<void> {
        await this.firestoreUsersCollection
            .doc(userId).collection(this.firestoreCollectionName)
            .doc(chatId).collection(this.firestoreMessagesSubcollectionName)
            .doc(messageId).delete();
    }

    static async messageQuery(
        queryKey: UserChatMessageQueryKey,
        args: UserChatMessageQueryArgs[UserChatMessageQueryKey],
    ): Promise<ChatMessage[]> {
        const query = await this.buildMessageQuery(queryKey, args);
        const querySnapshot = await query.get();
        return querySnapshot.docs.map((doc) => ChatMessage.fromJSON(doc.data() as Record<string, unknown>));
    }

    private static async buildMessageQuery(
        queryKey: UserChatMessageQueryKey,
        args: UserChatMessageQueryArgs[UserChatMessageQueryKey]): Promise<FirebaseFirestore.Query<FirebaseFirestore.DocumentData>> {
        switch (queryKey) {
            case UserChatMessageQueryKey.All: {
                const baseQuery = this.firestoreUsersCollection
                    .doc(args.userId)
                    .collection(this.firestoreCollectionName)
                    .doc(args.chatId)
                    .collection(this.firestoreMessagesSubcollectionName)
                    .orderBy("createdAt", "desc");
                if (args.limit) {
                    return baseQuery.limit(args.limit);
                }
                return baseQuery.orderBy("id", "asc");
            }
            default:
                throw SemurEngineErrorBuilder.internalError("Invalid chat message query key for users");
        }
    }
}
